### PR TITLE
Additional regression tests for issues that have been fixed previously

### LINKED ISSUE #10295: HasQueryFilter doesn't support "contains" in List or Array
The method HasQueryFilter doesn't support "contains" and evaluates the Query in memory instead of generating an SQL statement like 

```
SELECT [x].AccountId, [x].Name,  [x].TenantId FROM Account as x where [x].TenantId in ( tenantId_0, ...,  tenantId_n)
```

Log Output:

```
Compiling query model: 
'from Account x in DbSet<Account>
select new <>f__AnonymousType0<Guid, string, Guid>(
    [x].AccountId, 
    [x].Name, 
    [x].TenantId
)'

The LINQ expression 'where ___tenantIds_0.Contains([f].TenantId)' could not be translated and will be evaluated locally.
```

### Steps to reproduce
```c#
public class Account{
	public Guid AccountId {get;set}
	public string Name {get;set;}
	
	public Guid TenantId {get;set;}
}

public class DemoContext : DbContext
{
	private readonly List<Guid> _tenantIds;

	public virtual DbSet<Account> Accounts { get; set; }

	public DemoContext(DbContextOptions<DemoContext> options, IEnumerable<Guid> tenantIds)
		: base(options)
	{
		_tenantIds = tenantIds.Where(x => x.HasValue).ToList();
	}

	protected override void OnModelCreating(ModelBuilder modelBuilder)
	{
		modelBuilder.Entity<Account>().HasQueryFilter(e => _tenantIds.Contains(e.TenantId));
	}
}
```

Used Query

```c#
var data = context.Accounts
		.Select(x => new { x.AccountId, x.Name, x.TenantId })
		.ToList();
```

If you use the model without HasQueryFilter

```c#
 var data = context.Accounts
                    .Where(x => tenantIds.Contains(x.TenantId))
                    .Select(x => new { x.AccountId, x.Name, x.TenantId})
                    .ToList();
```

the generated Query looks like

```
SELECT [x].[AccountId], [x].[NAME], [x].[TenantId]
FROM [ng].[Accounts] AS [x]
WHERE [x].[TenantId] IN ('524996ea-5e4e-e611-80bd-0050569e091a', 'd5c43926-d830-e711-80c1-0050569e091a')
```

and works like expected.

### Further technical details
EF Core version: 2.0.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system:  Windows Server 2016
IDE: Visual Studio 2017

### LINKED ISSUE #12453: to put Take() in the different place will get different SQL
If we put "Take()" in the different place, we will get different SQL.

### Steps to reproduce
1. Just create a basic EF Core console project,
2. Create some POCO class and DBContext class, for example "CustomerOrderContext".
3. Do the following query/project.
4. Notice the different `Take(1)` method call.
5. Verify the log output.

## 1
```c#
using (CustomerOrderContext db = new CustomerOrderContext())
{
    var aa = db.Customers.Select(e => new Customer
    {
        FirstName = e.FirstName,
        LastName = e.LastName,
        UserName = e.UserName
    })
    .Take(1)
    .Select(e => new CusomterDto { FullName = e.FirstName + " " + e.LastName });
    foreach (var a in aa)
    {
        Console.WriteLine(a.FullName);
    }
}
```

If output the log, I can get the following information:
image

Be noted that it retrieves the "UserName" property, however that property doesn't used in the later.

## 2
```c#
using (CustomerOrderContext db = new CustomerOrderContext())
{
    var aa = db.Customers.Select(e => new Customer
    {
        FirstName = e.FirstName,
        LastName = e.LastName,
        UserName = e.UserName
    })    
    .Select(e => new CusomterDto { FullName = e.FirstName + " " + e.LastName })
    .Take(1);
    foreach (var a in aa)
    {
        Console.WriteLine(a.FullName);
    }
}
```
If output the log, I can get:
```C#
SELECT TOP(@__p_0) ([e].[FirstName] + N' ') + [e].[LastName] AS [FullName]
      FROM [Customers] AS [e]
```

Be noted that "UserName" property doesn't retrieve from the DB.

So, that's by design or did I make some mistaken or missed something? Thanks.

### LINKED ISSUE #13216: Where condition missing on Left join grouped sub query
I want left join twice, first is a grouped sub query, second is a table.

The major translation works well.

But the condition has some problem, if LINQ condtion constains the grouped query part, the translation will not translate condition part.

### Steps to reproduce
Include a complete code listing (or project/solution) that we can run to reproduce the issue.

Partial code listings, or multiple fragments of code, will slow down our response or cause us to push the issue back to you to provide code to reproduce the issue.

#### Tables
**Users**
userid, username
**SupperFCWacher**
id, userid
**FCWacher**
id , userid, dealerid

#### LINQ
```c#
var data = from u in _dbContext.Users
                       join s in _dbContext.SupperFCWachers
                       on u.Id equals s.UserId
                       into su
                       from m in su.DefaultIfEmpty()

                       join wg in (from w in _dbContext.FCWachers group w by w.UserId into g select new { UserId = g.Key, Count = g.Count() }
                       )
                       on u.Id equals wg.UserId
                       into wu
                       from n in wu.DefaultIfEmpty()
                       where m.Id != null || n.Count > 0
                       select new FCWacherSummaryModel()
                       {
                           UserId = u.Id,
                           UserName = u.UserName,
                           Description = m == null ? "" : "All Dealers"
                       };
```
#### Translated SQL
```SQL
SELECT [t].[UserId], [t].[Count], [s].[Id], [u].[UserID] AS [UserId0], [u].[UserName], CASE
    WHEN [s].[Id] IS NULL
    THEN N'' ELSE N'All Dealers'
END AS [Description]
FROM [UserDefinition] AS [u]
LEFT JOIN [SupperFCWacher] AS [s] ON [u].[UserID] = [s].[UserId]
LEFT JOIN (
    SELECT [w].[UserId], COUNT(*) AS [Count]
    FROM [FCWacher] AS [w]
    GROUP BY [w].[UserId]
) AS [t] ON [u].[UserID] = [t].[UserId]
--where s.id is not null or t.Count>0   --!!this condition will not be translated.
```
### Another Issue

If FCWacher inherit SupperFCWacher, the translated sql will be error.

### Further technical details
EF Core version: (2.1.1)
Database Provider: (Microsoft.EntityFrameworkCore.SqlServer)
Operating system: 
IDE: (e.g. Visual Studio 2017 15.4)

### LINKED ISSUE #13550: Join returns more columns then it should
When using join query returns more columns then it should.

### Steps to reproduce
This query:
```c#
            var temp = await this.Set<DbInvoiceState>().Join(
                    this.Set<DbInvoiceState>().GroupBy(t => t.InvoiceId).Select(t => new { t.Key, MaxInserted = t.Max(t2 => t2.InsertedDateTime) }),
                    t => new
                    {
                        t.InvoiceId,
                        t.InsertedDateTime,
                        states = states.Contains(t.DocumentState)
                    },
                    t => new
                    {
                        InvoiceId = t.Key,
                        InsertedDateTime = t.MaxInserted,
                        states = true
                    },
                    (t, t2) => t2.Key)
                .ToListAsync();
```
executes this query:
```
SELECT [t1].[Key], [t1].[MaxInserted]   //<--------- why this column?
FROM [Postman].[Invoice-States] AS [t]
INNER JOIN (
    SELECT [t0].[InvoiceId] AS [Key], MAX([t0].[InsertedDateTime]) AS [MaxInserted]
    FROM [Postman].[Invoice-States] AS [t0]
    GROUP BY [t0].[InvoiceId]
) AS [t1] ON (([t].[InvoiceId] = [t1].[Key]) AND ([t].[InsertedDateTime] = [t1].[MaxInserted])) AND [t].[DocumentState] IN (CAST(11 AS tinyint), CAST(12 AS tinyint), CAST(13 AS tinyint))
```
but it should be without `[t1].[MaxInserted`:
```
SELECT [t1].[Key]
FROM [Postman].[Invoice-States] AS [t]
INNER JOIN (
    SELECT [t0].[InvoiceId] AS [Key], MAX([t0].[InsertedDateTime]) AS [MaxInserted]
    FROM [Postman].[Invoice-States] AS [t0]
    GROUP BY [t0].[InvoiceId]
) AS [t1] ON (([t].[InvoiceId] = [t1].[Key]) AND ([t].[InsertedDateTime] = [t1].[MaxInserted])) AND [t].[DocumentState] IN (CAST(11 AS tinyint), CAST(12 AS tinyint), CAST(13 AS tinyint))
```

### Further technical details
EF Core version: 2.1.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Win-10-64bit
IDE: (Visual Studio 2017 15.8.3)

### Also tried
this query
```c#
            var temp = await this.Set<DbInvoiceState>().Join(
                    this.Set<DbInvoiceState>().GroupBy(t => t.InvoiceId).Select(t => new { t.Key, MaxInserted = t.Max(t2 => t2.InsertedDateTime) }),
                    t => new
                    {
                        t.InvoiceId,
                        t.InsertedDateTime,
                        states = states.Contains(t.DocumentState)
                    },
                    t => new
                    {
                        InvoiceId = t.Key,
                        InsertedDateTime = t.MaxInserted,
                        states = true
                    },
                    (t, t2) => t.InvoiceId)       //   <--------   changed
                .ToListAsync();
```
but nothing is chaged: `SELECT [t1].[Key], [t1].[MaxInserted], [t].[InvoiceId]`

### LINKED ISSUE #13712: sql server does not support correlated subquery in Take(n).Tolist()
I have a simple query with projection and correlated subquery and I have expect one query for query not N+1. I have to say that this code generates query that I expect with EF6.
and I know that query optimization added in 2.1 but I can not figure out this behavior.

### Steps to reproduce
different queries  are generate for each record for post.PostDisableCoins.ToList(). 

```c#
source.Select(post => new PostProject
            {
                
                PostDisableCoins = post.PostDisableCoins.ToList(),
                WarStartTime = post.WarStartTime,
                WarEndTime = post.WarEndTime,
                WarWinner = post.WarWinner,
                WarDeclarer = post.WarDeclarer
            });
```

### Further technical details
EF Core version: 2.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: 
IDE:  Visual Studio 2017 15.8.7

### LINKED ISSUE #13977: Column aliases ignored when distinct is used
Describe what is not working as expected.

Consider the following query:

```C#
var efforts = await (from se in _context.StudentEfforts
  join te in _context.TeacherEfforts on se.ReportingPeriodClassId equals te.ReportingPeriodClassId
  join ef in _context.Efforts on se.EffortId equals ef.Id
  join tt in _context.Teachers on te.TeacherId equals tt.Id
  join rpc in _context.ReportingPeriodClasses on se.ReportingPeriodClassId equals rpc.Id
  join cc in _context.Classes on rpc.ClassId equals cc.Id
  where rpc.ReportingPeriodId == periodId 
    && se.StudentId == studentId
  select new {
      Teacher = tt,
      ClassId = cc.Id,
      ClassName = cc.Name,
      Effort = ef.Name,       
      EffortId = ef.Id,
      TeacherGraded = te.Grade,
      StudentGraded = se.Grade
  }).Distinct().GroupBy(ef => ef.ClassId).ToListAsync();
```

Will generate the following SQL Code in which the aliases are ignored in the outer select, causing an invalid column name error.

```sql
exec sp_executesql N'
SELECT [t].[Id], [t].[Code], [t].[DateOfBirth], [t].[Email], [t].[FirstName], [t].[Gender], [t].[GenderId], [t].[IdentityId], [t].[LastName], [t].[MiddleName], [t].[Mobile], [t].[Title], [t].[ClassId], [t].[ClassName], [t].[Effort], [t].[EffortId], [t].[TeacherGraded], [t].[StudentGraded], [t].[Id], [t].[Name], [t].[Name], [t].[Id], [t].[Grade], [t].[Grade]
FROM (
    SELECT DISTINCT [tt].[Id], [tt].[Code], [tt].[DateOfBirth], [tt].[Email], [tt].[FirstName], [tt].[Gender], [tt].[GenderId], [tt].[IdentityId], [tt].[LastName], [tt].[MiddleName], [tt].[Mobile], [tt].[Title], [cc].[Id] AS [ClassId], [cc].[Name] AS [ClassName], [ef].[Name] AS [Effort], [ef].[Id] AS [EffortId], [te].[Grade] AS [TeacherGraded], [se].[Grade] AS [StudentGraded]
    FROM [StudentEfforts] AS [se]
    INNER JOIN [TeacherEfforts] AS [te] ON [se].[ReportingPeriodClassId] = [te].[ReportingPeriodClassId]
    INNER JOIN [Efforts] AS [ef] ON [se].[EffortId] = [ef].[Id]
    INNER JOIN [Teachers] AS [tt] ON [te].[TeacherId] = [tt].[Id]
    INNER JOIN [ReportingPeriodClasses] AS [rpc] ON [se].[ReportingPeriodClassId] = [rpc].[Id]
    INNER JOIN [Classes] AS [cc] ON [rpc].[ClassId] = [cc].[Id]
    WHERE ([rpc].[ReportingPeriodId] = @__periodId_0) AND ([se].[StudentId] = @__studentId_1)
) AS [t]',N'@__periodId_0 int,@__studentId_1 int',@__periodId_0=2,@__studentId_1=610

```
However removing the Distinct() from the query generates working SQL code

### LINKED ISSUE #15302: ArgumentNullException on query that worked in EF 6
Note sure if this issue is classed as the same as the currently open one here [Issue 12603](https://github.com/aspnet/EntityFrameworkCore/issues/12603)

So I have converted an application from wpf and EF6 to .net core 3 and EF core and
I have a query that seemed to work fine in EF 6 but gives me a ArgumentNullException
in EF core. 
The annoying thing is that I tried creating a new app and a new database, stripping out tables and columns that are not required by the query, just so I could try and recreate the issue. Query works fine in this new recreated app, although the database only contains 2 records in each table.

Here is the query code.

```c#
var items2 = DbContext.UutResult.Where(x =>
                x.StartDateTime != null && (x.StartDateTime.Value.Date >= startDate &&
                                            x.StartDateTime.Value.Date <= endDate &&
                                            x.StationId == testerId)).Select(x => new TestResultItem()
            {
                TesterId = (int) x.TestSocketIndex,
                TesterType = x.StationId,
                TestDateTime = (DateTime) x.StartDateTime,
                TestResult = x.UutStatus,
                PanelBarcode = x.BatchSerialNumber,
                UutBarcode = x.UutStatus,
                TestTimeSec = (double) x.ExecutionTime,
                TestSteps = x.StepResult.Where(y => y.StepGroup == "Main").Select(y => new TestResultStepItem()
                {
                    StepName = y.StepName,
                    Report = y.ReportText,
                    TestResult = y.Status,
                    UpperLimit = y.StepNumericlimit1.FirstOrDefault().StepNumericlimit2.FirstOrDefault().HighLimit.ToString(),
                    OrderId = (int)y.OrderNumber
                }).ToList()
            }).ToList();
```

The exception is on this line, because if I comment it out then the query runs fine.

```
UpperLimit = y.StepNumericlimit1.FirstOrDefault().StepNumericlimit2.FirstOrDefault().HighLimit.ToString()
```

I assumed it was because StepNumericlimit1 could be null or StepNumericlimit2 could be null, but in the new dummy database I made sure these null conditions were there i.e. I made sure there was 1 StepNumericlimit1 with 0 child StepNumericlimit2, and it still worked.

```
Exception message: System.ArgumentNullException: 'Value cannot be null. Parameter name: source'
Stack trace: at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument) at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable1 source, Boolean& found)
     at lambda_method(Closure , QueryContext , TransparentIdentifier2 ) at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ProjectionShaper.TypedProjectionShaper3.Shape(QueryContext queryContext, ValueBuffer& valueBuffer)
     at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.ProjectionShaper.TypedProjectionShaper3.Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.IShaper.Shape(QueryContext queryContext, ValueBuffer& valueBuffer) at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable1.Enumerator.BufferlessMoveNext(DbContext _, Boolean buffer)
     at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func3 operation, Func3 verifySucceeded)
     at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable1.Enumerator.MoveNext() at Microsoft.EntityFrameworkCore.Query.Internal.QueryBuffer.CorrelateSubquery[TInner,TOut,TCollection](Int32 correlatedCollectionId, INavigation navigation, Func2 resultCollectionFactory, MaterializedAnonymousObject& outerKey, Boolean tracking, Func1 correlatedCollectionFactory, Func3 correlationPredicate)
     at lambda_method(Closure , ValueBuffer )
     at System.Linq.Enumerable.SelectEnumerableIterator2.MoveNext() at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor1.EnumeratorExceptionInterceptor.MoveNext()
     at System.Collections.Generic.List1..ctor(IEnumerable1 collection)
     at System.Linq.Enumerable.ToList[TSource](IEnumerable1 source) at EPQT.ServiceLayer.TesterService.GetTestData(DateTime startDate, DateTime endDate, String testerId, String productFilter) in C:\Visual Studio 2017\Projects\Extractor\Core\EPQT.ServiceLayer\TesterService.cs:line 120 at EPQT.Modules.TesterQuery.Views.TestQueryViewModel.<>c__DisplayClass54_0.b__0() in C:\Visual Studio 2017\Projects\Extractor\Modules\EPQT.Modules.TesterQuery\Views\TestQueryViewModel.cs:line 218 at System.Threading.Tasks.Task.InnerInvoke() at System.Threading.Tasks.Task.<>c.<.cctor>b__277_0(Object obj) at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
```

### Further technical details
EF Core version: 2.2.4
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual studio 2019

### EF Code files - scaffold generated files.
[EFCode_NonWorking.zip](https://github.com/aspnet/EntityFrameworkCore/files/3062317/EFCode_NonWorking.zip)
[EFCode_Working.zip](https://github.com/aspnet/EntityFrameworkCore/files/3062336/EFCode_Working.zip)

### LINKED ISSUE #17735: Query: Translate List<T>.Count to server side
We can convert it to Count() function. Currently it client evals materializing whole list.

### PR DESCRIPTION
**Pull request by maumar**
Resolves #10295
Resolves #12453
Resolves #13216
Resolves #13550
Resolves #13712
Resolves #13977
Resolves #15302
Resolves #17735

### CODE REVIEW

File: test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs [hunk 1]
```diff
@@ -1174,6 +1174,33 @@ public override Task Projecting_Length_of_a_string_property_after_FirstOrDefault
...
+        public override Task Projecting_count_of_navigation_which_is_generic_list(bool async)
+        {
+            return base.Projecting_count_of_navigation_which_is_generic_list(async);
+        }
+
+
+
+
+
+
```
Comment (smitpatel) [inline test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs:1203]:
```suggestion

```

File: test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs [hunk 2]
```diff
@@ -1174,6 +1174,33 @@ public override Task Projecting_Length_of_a_string_property_after_FirstOrDefault
             return base.Projecting_Length_of_a_string_property_after_FirstOrDefault_on_correlated_collection(async);
         }
 
+        // skip without take
+        public override Task Projection_skip_projection_doesnt_project_intermittent_column(bool async)
```
Comment (smitpatel) [inline test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs:1178]: Can you verify exception instead?

File: test/EFCore.Specification.Tests/Query/NorthwindSelectQueryTestBase.cs [hunk 3]
```diff
@@ -1860,5 +1861,111 @@ public virtual Task Projecting_Length_of_a_string_property_after_FirstOrDefault_
...
+                    .OrderBy(c => c.CustomerID)
+                    .Select(c => new { c.CustomerID, c.City, c.CompanyName })
+                    .Skip(7)
+                    .Select(x => new { Aggregate = x.CustomerID + " " + x.City }),
+                assertOrder: true);
+        }
+
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Projection_Distinct_projection_contains_intermittent_column(bool async)
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/NorthwindSelectQueryTestBase.cs:1942]: Can you rename this test? It does not project intermittent column in final projection but uses a subquery for distinct.

File: test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs:1178]: verify SQL for other 2 test which are not overriden here.

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs [hunk 4]
```diff
@@ -7898,6 +7898,49 @@ public virtual async Task Accessing_derived_property_using_hard_and_soft_cast(bo
                 ss => ss.Set<LocustLeader>().Where(ll => ll is LocustCommander && (ll as LocustCommander).HighCommandId != 0));
         }
 
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Cast_to_derived_followed_by_include_and_FirstOrDefault(bool async)
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs:7903]: Is there a linking issue for this?

File: test/EFCore.SqlServer.FunctionalTests/Query/TPTGearsOfWarQuerySqlServerTest.cs [hunk 5]
```diff
@@ -8342,6 +8342,33 @@ public override async Task Accessing_derived_property_using_hard_and_soft_cast(b
 WHERE [l0].[Name] IS NOT NULL AND (([l0].[HighCommandId] <> 0) OR [l0].[HighCommandId] IS NULL)");
         }
 
+        public override async Task Cast_to_derived_followed_by_include_and_FirstOrDefault(bool async)
+        {
+            await base.Cast_to_derived_followed_by_include_and_FirstOrDefault(async);
+
+            AssertSql(
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/TPTGearsOfWarQuerySqlServerTest.cs:8349]: ?

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs [hunk 6]
```diff
@@ -5930,6 +5930,123 @@ ELSE CAST(0 AS bit)
...
+    SELECT [l0].[Id] % 3 AS [c], COALESCE(SUM([l0].[Id]), 0) AS [c0]
+    FROM [LevelTwo] AS [l0]
+    GROUP BY [l0].[Id] % 3
+) AS [t] ON ([l].[Id] = [t].[c]) AND (CAST(1 AS bit) = CASE
+    WHEN [t].[c0] > 10 THEN CAST(1 AS bit)
+    ELSE CAST(0 AS bit)
+END)");
+        }
+
+
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs:5950]:
```suggestion

```

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs [hunk 7]
```diff
@@ -5930,6 +5930,123 @@ ELSE CAST(0 AS bit)
...
+        public override async Task Projecting_columns_with_same_name_from_different_entities_making_sure_aliasing_works_after_Distinct(bool async)
+        {
+            await base.Projecting_columns_with_same_name_from_different_entities_making_sure_aliasing_works_after_Distinct(async);
+
+            AssertSql(
+                @"@__p_0='10'
+
+SELECT [t].[Id] AS [Foo], [t].[Id0] AS [Bar], [t].[Id1] AS [Baz]
+FROM (
+    SELECT DISTINCT TOP(@__p_0) [l].[Id], [l0].[Id] AS [Id0], [l1].[Id] AS [Id1], [l].[Name], [l0].[Name] AS [Name0]
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs:6043]: Distinct should be applying alias from the projection before Distinct. The aliases here should be ID1-3 Name1-2

**Review by smitpatel — APPROVED** [review]

File: test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs (hunk 4 above)
Comment (maumar) [reply test/EFCore.Specification.Tests/Query/GearsOfWarQueryTestBase.cs:7903]: part of regression tests for #22692

File: test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs (hunk 7 above)
Comment (maumar) [reply test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs:6043]: a bug then?

Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/ComplexNavigationsQuerySqlServerTest.cs:6042]: Not the original bug but a different bug. Just file a new issue with link to this test. I will look into it.

File: test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs [hunk 8]
```diff
@@ -1174,6 +1174,40 @@ public override Task Projecting_Length_of_a_string_property_after_FirstOrDefault
             return base.Projecting_Length_of_a_string_property_after_FirstOrDefault_on_correlated_collection(async);
         }
 
+        public override async Task Projection_take_predicate_projection(bool async)
+        {
+            await base.Projection_take_predicate_projection(async);
+
+            AssertSql(
```
Comment (smitpatel) [inline test/EFCore.Cosmos.FunctionalTests/Query/NorthwindSelectQueryCosmosTest.cs:1181]: No sql generated here?

**Pull request merged**